[github] Specialized errors

chantra opened the pull request:
[github] Use specialized error

Most, if not all errors, are generated by, or wrapped by, anyhow!.
This causes all errors to look the same from the caller vantage point, preventing
the caller from possibly handling them. For instance, when the API is
rate limited, returning an error that indicates this is a RateLimiting
error, and providing a simple way to know until when we are being rate
limited is valuable. At the moment, one needs to extract the message from the
anyhow error, do some regex magic to detect if this is being rate
limited or not, and extract until when the rate limitation applies.

This change introduces the use of the crate `thiserror` and ports the
github crate to use it.
All crates will now return a `ClientResult` instead of a
`anyhow::Result`. For crates ported to `thiserror`, Client result is an
alias to `Result<T, ClientError>`, where `ClientError` is the enum
defined using `thiserror` to populate variants.
If a crate is not onboarded in `thiserror`, `ClientResult` is a simple
alias to `anyhow::Result`.

chantra commented:
please ignore the first commit ([7947203](https://github.com/oxidecomputer/third-party-api-clients/pull/57/commits/7947203652fe594edda20eb2b7f0d3a3afd12f12)), which is what get spit out when I run `make github` on top of the current main branch.

Actual change is in [4eb403d](https://github.com/oxidecomputer/third-party-api-clients/pull/57/commits/4eb403dea3a4f38417998b6302c6153635e4c5a0).


I have only rebuild `github` for now, but tested this also on other projects by running `make`. It generates too much noise to be valuable here. Once I have gotten feedback, I will be more than happy to provide a change which would include everything.

https://gist.github.com/chantra/f1b450ce34fec60061048073df3d7195 would be the generated remaining diff.

chantra commented:
actually ported all projects to the specialized errors. The interesting commits are the 3 last ones. The first one just rebuilds them all against my setup which generates a bunch of noisy change.

augustuswm commented:
Thanks for putting this all together. I'm definitely on board with the premise of returning specific errors instead of an `anyhow` error. Give me a little to read through this.

augustuswm commented on generator/src/main.rs:
```diff
@@ -2350,7 +2350,56 @@ fn gen(
+/// Errors returned by the client
+#[derive(Debug, Error)]
+pub enum ClientError {
+    // Github only
```
When generating this enum, it shouldn't contain variants that we know won't be used. Specifically this would affect `RateLimited` and `EmptyRefreshToken`, though I haven't looked through all of them yet.

Currently the generator uses the `proper_name` variable in `gen` to differentiate.

augustuswm replied:
As long as the known client specific variants are split off that should be sufficient for now (i.e. the GitHub ones). Keeping the `#[from]` derived variants is fine. Categorizing the clients into an enum may work, but I'm not sure if it can be used generally, as there are a decent number of client specific branches.

Overall this is looking good.

chantra commented:
@augustuswm let me know how that look like. I can update this PR with only the 2 meaningful commits (vs the rebuild commits) and let CI rebuild upon release.

The conflict shown is because #58 change was not rebuilt.

augustuswm commented:
Yeah, if you don't mind reducing it down to only the non-generated files, that would probably be cleaner for a commit. The builds do add a lot of noise. `main` should now be updated with #58 and rebuild clients.

chantra commented:
Thanks. I stripped the 2 extra commits. So only relevant code change is left now.

augustuswm commented on generator/src/main.rs:
```diff
@@ -2348,9 +2368,77 @@ fn gen(
+    },
+    /// Generic errors returned by the API.
+    #[error(transparent)]
+    GenericError(#[from] anyhow::Error),
```
Can we limit this (`GenericError`) variant to the Google Drive client only. Looks like it is the only location it is needed. (And eventually will be replaced with a specific error)

chantra replied:
Fair point. Let me port those remaining errors to something more specific and do without anyhow::Error:
```
./google/drive/src/traits.rs:use anyhow::anyhow;
./google/drive/src/traits.rs:            None => return Err(ClientError::GenericError(anyhow!("No Location header"))),
./google/drive/src/traits.rs:        Err(ClientError::GenericError(anyhow!(
./google/sheets/src/traits.rs:use anyhow::Result;
./sendgrid/src/traits.rs:use anyhow::{anyhow, Result};
./sendgrid/src/traits.rs:            s => Err(anyhow!("received response status: {:?}", s)),
```

augustuswm commented:
One last edit before merging this. We may iterate on the error variants, but overall this makes sense and it would be good to start building off of it.

chantra commented:
> One last edit before merging this. We may iterate on the error variants, but overall this makes sense and it would be good to start building off of it.

ok, addressed comments in https://github.com/chantra/oxidecomputer-third-party-api-clients/commit/2b0b71fa4a569dae0b7000382c8dba8babcacd6e

anyhow has been removed from the projects' cargo files as part of this.

chantra commented:
Thanks @augustuswm 
Looks like I forgot the httpcache use case. Will fix forward in a couple of hours. Do you know if there is any other features I should test for other clients?

chantra commented:
should be fixed in #63 . Sorry for the trouble.

augustuswm commented:
Thanks! All good, I'll get that merged.
